### PR TITLE
Add `requires_initialization` property to trust regions

### PR DESCRIPTION
**Pull request by khurram-ghani**
**Related issue(s)/PRs:** resolves re-init issue discussed in #825 

## Summary
This PR adds a property to trust-regions, allowing a rule to determine if a region needs to be re-initialized. The rule can then call the `initialize` method directly on the region. Previously this decision was made internally by the regions. This resulted in an issue for product spaces where the re-init of sub-regions wasn't coordinated. This PR resolves that issue.

**Fully backwards compatible:** no
There is a minor incompatibility that the regions are now required to set `self._initialized` to `True` in their `initialize` method. This instance variable starts off as `False` and should be set to `True` in any derived `initialize` methods.

## PR checklist

- [X] The quality checks are all passing
- [X] The bug case / new feature is covered by tests
- [X] Any new features are well-documented (in docstrings or notebooks)

### CODE REVIEW

File: tests/unit/acquisition/test_rule.py
Comment (hstojic) [inline tests/unit/acquisition/test_rule.py]: move the comments to the docstrings of the test functions?

**Review by hstojic — APPROVED** [review]
it looks good to me, this should resolve the issue with product search spaces!
but lets wait for @uri-granta 's approval as well :)

File: trieste/acquisition/rule.py
Comment (uri-granta) [inline trieste/acquisition/rule.py]: (suggestion)

```suggestion
    def requires_initialization(self) -> bool:
```

File: trieste/acquisition/rule.py
Comment (uri-granta) [inline trieste/acquisition/rule.py]: Maybe expand here a bit?

```suggestion
        """Return `True` if the search space needs to be re-initialized with the latest datasets, and `False` if it can be just updated."""
```

File: trieste/acquisition/rule.py
Comment (uri-granta) [inline trieste/acquisition/rule.py]: Do we want to move `_initialized` into the UpdatableTrustRegion base class rather than implementing it each time? We could then default to returning `self._initialized`. (If you do this, then you can also remove the "initialization" in `FixedPointTrustRegionDiscrete.__init__`.)

**Review by uri-granta — APPROVED** [review]
Looks good. A couple of suggestions/questions.

File: trieste/acquisition/rule.py
Comment (khurram-ghani) [reply trieste/acquisition/rule.py]: Yes, makes sense. I thought about this before, but didn't go ahead with it. Now that you mention it ...

I have made the change, though I kept the "initialization" in `FixedPointTrustRegionDiscrete.__init__`. Safer to set the points to something (rather than `None`), and random init is the most obvious.

Comment (khurram-ghani) [reply trieste/acquisition/rule.py]: @uri-granta note though that this change adds a minor incompatibility, explained in the description:

> There is a minor incompatibility that the regions are now required to set `self._initialized` to `True` in their `initialize` method. This instance variable starts off as `False` and should be set to `True` in any derived `initialize` methods.

Thoughts?

Comment (uri-granta) [reply trieste/acquisition/rule.py]: I think it's fine, though I would document this in the `initialize` abstractmethod docstring. Alternatively, you could make `initialize` not abstract and default to just setting `self._initialized` to `True`, and call it via `super()` everywhere. Though that adds more code for no obvious benifit.

Comment (khurram-ghani) [reply trieste/acquisition/rule.py]: Thanks. I have added a comment in the docstring.

I agree that making `initialize` not abstract would be of limited benefit, plus it would not fix the backwards incompatibility, since existing classes would still need to be updated to call `super()...`.

**Pull request merged**
